geps/gep-1748.md

```diff
@@ -0,0 +1,140 @@
+“/east” to the store-east ServiceImport. Requests that don’t match either of
+these paths will be routed to the “store” ServiceImport which represents a
+superset of the “store-west” and “store-east” ServiceImports.
+
+```yaml
+{% include 'standard/multicluster/httproute-location.yaml' %}
+```
+
+#### Advanced Routing With ServiceImports
+
+All Routing capabilities in Gateway API should apply equally whether the backend
+is a Service or ServiceImport. For example, when routing to a system with
+multiple read replicas, it could be beneficial to route requests based on HTTP
+Method. In the following example, requests with POST, PUT, and DELETE methods
+are routed to `api-primary` while the rest are routed to `api-replicas`:
+
+```yaml
+{% include 'standard/multicluster/httproute-method.yaml' %}
+```
+
+#### Routing to Both Services and ServiceImports
+
+There are some situations where it will be useful to split traffic between a
+Service and ServiceImport. In the following example, 90% of traffic would go to
+endpoints attached to the cluster-local "store" Service, and the remaining 10%
+would go to endpoints attached to the Multi-Cluster "store-global" Service:
+
+```yaml
+{% include 'standard/multicluster/httproute-hybrid.yaml' %}
+```
+
+
+#### Cross-Namespace References with ReferenceGrant
+
+It is possible to use ReferenceGrant to enable cross-namespace references to
+ServiceImports. For example, the following HTTPRoute would forward traffic to
+endpoints attached to the “bar” Service in the “bar” namespace:
+
+```yaml
+{% include 'standard/multicluster/httproute-referencegrant.yaml' %}
+```
+
+### Mesh: ServiceImport as Parent
+
+In some cases, you may want to override traffic destined for a Multi-Cluster
+Service with a mesh. As part of the broader GAMMA initiative, ServiceImport can
+be used in the same way that Service is used as a ParentRef. When a Service is
+specified as a parent, meshes will intercept traffic destined for the ClusterIP
+of the Service and apply any policies or routing decisions defined by the Route.
+Similarly, when a ServiceImport is specified as a parent, meshes will intercept
+traffic destined for the ClusterSetIP and apply any policies or routing
+decisions defined by the Route. In the following example, the mesh would
+intercept traffic destined for the store ClusterSetIP and redirect it to
+endpoints for the local Service within the same cluster.
+
+```yaml
+{% include 'standard/multicluster/httproute-gamma.yaml' %}
+```
+
+### Services vs ServiceImports
+
+It is important that all implementations provide a consistent experience. That
+means that references to Services MUST always be interpreted as references to
+endpoints within the same cluster for that Service. References to ServiceImports
+MUST be interpreted as references to the full set of endpoints attached to that
+ServiceImport. In practice, that means that users should use “Service” when they
```

Comment:
This is the biggest ⚠️ for me, and is a bit tricky at multiple levels:
- Reading this the first time, I had assumed `"full set of endpoints"` was intending to strictly define _the set of endpoints expected to be attached to a ServiceImport_ rather than just how they should be consumed, which would have been a significant expectation above the guidance in [KEP-1645](https://github.com/kubernetes/enhancements/tree/master/keps/sig-multicluster/1645-multi-cluster-services-api#importing-services):
  > The `EndpointSlices` imported for a service are not guaranteed to exactly match the originally exported `EndpointSlices`, but each slice is guaranteed to map only to a single source cluster.

  I've heard guidance elsewhere that the _endpoints attached to a ServiceImport_ should be "representative" of the available multicluster endpoints (citation needed cc @JeremyOT) rather than necessarily the full set, which is important for allowing a ServiceImport to:
  - _directly_ expose only a subset of pod endpoints from other clusters, particularly in clustersets with a very large number of endpoints for a ServiceExport
  - _indirectly_ expose imported services through a transit gateway of some sort (rather than allowing direct pod <> pod connections across clusters), which may otherwise necessitate a permissive networking environment with egress and ingress allowed across a wide range of IPs and/or the assumption of a "flat" non-overlapping networking shared by all clusters (IIRC "reachability" of imported services is an assumption/requirement of MCS, but the exact mechanics of how to implement that are intentionally not proscribed)
- The phrase `"references to ... endpoints"` seems like it _might_ be understood as implying "direct" pod addressing, which could preclude using gateways between clusters as described above. Cloud provider MCS implementations do commonly enable this through VPCs, but some Gateway API implementations may want to reference off-cluster or multi-network services without relying on this substrate.
  - If this were clarified to more explicitly describe the contents of the EndpointSlices associated with a ServiceImport (whatever those may point to) this might be flexible enough.

There's some additional complexity around how cross-cluster "direct endpoint" routing might work for GAMMA, but I think that scope could be deferred if we can affirm that ServiceImport "indirect" routing through gateways in other clusters is a permissible implementation.

Reply:
Good catch, thanks @mikemorris! I think you're right that the MCS Spec just requires a representative set, not the full set. I also agree that it allows for gateways between clusters or other intermediate steps. I've tried to reword this section to avoid requiring either direct pod addressing or routing to the full set of endpoints from each cluster. Not sure I've got it quite right yet, so let me know if you can think of a better way to state this requirement.

Reply:
Resolving this for now, let me know if I missed anything @mikemorris